netket/machine.py

```diff
@@ -1,5 +1,256 @@
+import numpy as _np
 
 
 def MPSPeriodicDiagonal(hilbert, bond_dim, symperiod=-1):
     return MPSPeriodic(hilbert, bond_dim, diag=True, symperiod=symperiod)
+
+
+class CxxMachine(Machine):
+    def __init__(self, hilbert):
+        super(CxxMachine, self).__init__(hilbert)
+
+    def log_val(self, x):
+        r"""Returns the logarithm of the wave function at point x.
+
+        Subclasses should implement this function.
+
+        Args:
+            x: 1D vector of `float` of size `self.n_visible`.
+        """
+        raise NotImplementedError
+
+    def der_log(self, x):
+        r"""Returns the gradient of the logarithm of the wave function at point
+        x.
+
+        Subclasses should implement this function.
+
+        Args:
+            x: 1D vector of `float` of size `self.n_visible`.
+        """
+        raise NotImplementedError
+
+    def save(self, filename):
+        r"""Saves machine's state to file.
+
+        Subclasses should implement this function.
+
+        *Note:* the sequence of saving and loading, -- ``m.save(filename);
+        m.load(filename)``, -- should be an identity.
+
+        Args:
+            filename: name of the file.
+	"""
+        raise NotImplementedError
+
+    def load(self, filename):
+        r"""Loads machine's state from file.
+
+        *Note:* the sequence of saving and loading, -- ``m.save(filename);
+        m.load(filename)``, -- should be an identity.
+
+        Subclasses should implement this function.
+
+        Args:
+            filename: name of the file.
+	"""
+        raise NotImplementedError
+
+    def _is_holomorphic(self):
+        r"""Returns whether the wave function is holomorphic.
+        """
+        raise NotImplementedError
+
+    def _number_parameters(self):
+        r"""Returns the number of variational parameters in the machine.
+
+        Subclasses should implement this function, but to actually get the
+        number of parameters, use `self.n_par` attribute.
+        """
+        raise NotImplementedError
+
+    def _number_visible(self):
+        r"""Returns the number of "visible units" in the machine.
+
+        Subclasses should implement this function, but to actually get the
+        dimension of input vector, use `self.n_visible` attribute.
+        """
+        raise NotImplementedError
+
+    def _get_parameters(self):
+        r"""Returns the variational parameters as a 1D complex vector.
+
+        Subclasses should implement this function, but to actually access
+        the parameters, use `self.parameters` attribute.
+        """
+        raise NotImplementedError
+
+    def _set_parameters(self, p):
+        r"""Sets the variational parameters.
+
+        Subclasses should implement this function, but to actually access
+        the parameters, use `self.parameters` attribute.
+
+        Args:
+            p: New variational parameters as a 1D complex vector.
+        """
+        raise NotImplementedError
+
+
+class PyRbm(CxxMachine):
```

Comment:
Can you put this in another file? Thanks

Reply:
This should actually be in `Examples/` to show how to derive a machine from python?

Reply:
On second though.... How should the tests work then? PyRbm is currently tested with all other machines by Test/Machines/test_machine.py script.